### ISSUE TITLE
The order of weights is changed for array of layers inside sub classed model

### ISSUE THREAD

**Issue by RanjanBalappa** [tensorflow/tensorflow#41152]
The order of moving variance and moving mean weights(BatchNormalization layer) changes when array of layers is created inside sub classed model

````python
import tensorflow as tf
from tensorflow.keras.layers import *
print(tf.__version__)

class A(tf.keras.layers.Layer):
    def  __init__(self,  name):
      super(A, self).__init__(name=name)
      self.conv = Conv2D(32, 3)
      self.bn = BatchNormalization()  
    def call(self,  x):
      x = self.conv(x)
      x = self.bn(x)
      return x

class B(tf.keras.Model):
    def __init__(self):
        super(B, self).__init__()
        self.blocks = []
        for i in range(3):
          self.blocks.append(A(f'a{i}'))    
    
    def call(self, x):
        for block in self.blocks:
          x = block(x)
        return x

b = B()
b.build((1, 128, 128, 3))
print([weight.name for weight in b.weights])

#The output is as below
'a0/conv2d/kernel:0',
 'a0/conv2d/bias:0',
 'a0/batch_normalization/gamma:0',
 'a0/batch_normalization/beta:0',
 'a1/conv2d_1/kernel:0',
 'a1/conv2d_1/bias:0',
 'a1/batch_normalization_1/gamma:0',
 'a1/batch_normalization_1/beta:0',
 'a2/conv2d_2/kernel:0',
 'a2/conv2d_2/bias:0',
 'a2/batch_normalization_2/gamma:0',
 'a2/batch_normalization_2/beta:0',
 'a0/batch_normalization/moving_mean:0',
 'a0/batch_normalization/moving_variance:0',
 'a1/batch_normalization_1/moving_mean:0',
 'a1/batch_normalization_1/moving_variance:0',
 'a2/batch_normalization_2/moving_mean:0',
 'a2/batch_normalization_2/moving_variance:0'

**Comment by Saduf2019** [comment]
I am able to replicate the issue, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/ed97214ddbbded31a4e0b6b68b3cdba4/untitled266.ipynb).

**Comment by saikumarchalla** [comment]
Was able to reproduce the issue in TF 2.5. Please find the gist [here](https://colab.research.google.com/gist/saikumarchalla/3f8c9268584e2ac10fbe7c731fa963d2/untitled87.ipynb).

**Comment by SuryanarayanaY** [comment]
Hi @RanjanBalappa,
Sorry for the delayed response. Here is my observations on the output you mentioned for the given code.Each BatchNormalisation layer has two `trainable weights` and two `non_trainable weights` and hence 3-layers in given problem has total 6 non_trainable weights.Tensorflow arranged them in such a way that all trainable parameters grouped in sequence and non-trainable parameters grouped and stored at last in the array.If we see the output all the trainable parameters stored first layer wise in order and then non_trainable parameters.The weight updations and normalisation taken care by Tensorflow underhood.Please refer the attached gist [here](https://colab.research.google.com/gist/SuryanarayanaY/04d1e7183fad938c37c1f586a7db190b/41152.ipynb) to check trainable and non_trainable weights.

If you have observed any deviations from the intended behaviour due to the storage order of these weights please let us know.

Thankyou!
